Source: https://www.practicaldatascience.org/html/exercises/Exercise_dataframe.html <br>
Cheat Sheet: https://github.com/pandas-dev/pandas/blob/master/doc/cheatsheet/Pandas_Cheat_Sheet.pdf

In this exercise, we’re going to use data from the American Communities Survey (ACS) to study the relationship betwen educational attainment and wages. The ACS is a survey conducted by the United States Census Bureau (though it is not “The Census,” which is a counting of every person in the United States that takes place every 10 years) to measure numerous features of the US population. The data we will be working with includes about 100 variables from the 2017 ACS survey, and is a 10% sample of the ACS (which itself is a 1% sample of the US population, so we’re working with about a 0.1% sample of the United States).

This data comes from IPUMS, which provides a very useful tool for getting subsets of major survey datasets, not just from the US, but from government statistical agencies the world over.

This is real data, meaning that you are being provided the data as it is provided by IPUMS. Documentation for all variables used in this data can be found here (you can either search by variable name to figure out the meaning of a variable in this data, or search for something you want to see if a variable with the right name is in this data).

Within this data is information on both the educational background and current earnings of a representative sample of Americans. We will now use this data to estimate the labor-market returns to graduating high school and college, and to learn something about the meaning of an educational degree.

1) Data for these exercises can be found here. First, download US_ACS_2017_10pct_sample.dta.

2) Now import US_ACS_2017_10pct_sample.dta into a pandas DataFrame. This can be done with the command pd.read_stata, which will read in files created in the program Stata (and which uses the file suffix .dta). This is a format commonly used by social scientists.


# Download data

In [4]:
!curl -LJO https://github.com/nickeubank/MIDS_Data/raw/master/US_AmericanCommunitySurvey/US_ACS_2017_10pct_sample.dta

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   191  100   191    0     0     66      0  0:00:02  0:00:02 --:--:--    66
100 62.1M  100 62.1M    0     0  1020k      0  0:01:02  0:01:02 --:--:--  999k 6 62.1M    6 3862k    0     0   338k      0  0:03:08  0:00:11  0:02:57  925k.1M    8 5318k    0     0   428k      0  0:02:28  0:00:12  0:02:16 1105k5 62.1M   15 9846k    0     0   599k      0  0:01:46  0:00:16  0:01:30 1194k62.1M   17 10.5M    0     0   621k      0  0:01:42  0:00:17  0:01:25 1102k 62.1M   25 15.8M    0     0   658k      0  0:01:36  0:00:24  0:01:12  717k 23.6M    0     0   771k      0  0:01:22  0:00:31  0:00:51 1300k62.1M   93 58.2M    0     0  1017k      0  0:01:02  0:00:58  0:00:04  947k


In [5]:
!ls

DataFrames.ipynb             US_ACS_2017_10pct_sample.dta
Series.ipynb


# Load data to Pandas

In [12]:
import pandas as pd
df = pd.read_stata('US_ACS_2017_10pct_sample.dta')
df.shape

(319004, 104)

In [11]:
df.head()

year  datanum   serial      cbserial          numprec subsamp  hhwt  \
0  2017        1   177686  2.017001e+12                9      64    55   
1  2017        1  1200045  2.017001e+12                6      79    25   
2  2017        1    70831  2.017000e+12  1 person record      36    57   
3  2017        1   557128  2.017001e+12                2      10    98   
4  2017        1   614890  2.017001e+12                4      96    54   

                                   hhtype       cluster    adjust  ...  \
0  female householder, no husband present  2.017002e+12  1.011189  ...   
1       male householder, no wife present  2.017012e+12  1.011189  ...   
2          male householder, living alone  2.017001e+12  1.011189  ...   
3         married-couple family household  2.017006e+12  1.011189  ...   
4         married-couple family household  2.017006e+12  1.011189  ...   

   migcounty1                 migmet131 vetdisab                   diffrem  \
0           0  not in identifiable area      n/a                       n/a   
1           0  not in identifiable area      n/a   no cognitive difficulty   
2           0  not in identifiable area      n/a  has cognitive difficulty   
3           0  not in identifiable area      n/a   no cognitive difficulty   
4           0  not in identifiable area      n/a                       n/a   

                   diffphys                           diffmob diffcare  \
0                       n/a                               n/a      n/a   
1  no ambulatory difficulty  no independent living difficulty       no   
2  no ambulatory difficulty  no independent living difficulty       no   
3  no ambulatory difficulty  no independent living difficulty       no   
4                       n/a                               n/a      n/a   

                          diffsens  diffeye  diffhear  
0  no vision or hearing difficulty       no        no  
1  no vision or hearing difficulty       no        no  
2  no vision or hearing difficulty       no        no  
3  no vision or hearing difficulty       no        no  
4  no vision or hearing difficulty       no        no  

[5 rows x 104 columns]

In [13]:
for c in df.columns: print(c)

year
datanum
serial
cbserial
numprec
subsamp
hhwt
hhtype
cluster
adjust
cpi99
region
stateicp
statefip
countyicp
countyfip
metro
city
citypop
strata
gq
farm
ownershp
ownershpd
mortgage
mortgag2
mortamt1
mortamt2
respmode
pernum
cbpernum
perwt
slwt
famunit
sex
age
marst
birthyr
race
raced
hispan
hispand
bpl
bpld
citizen
yrnatur
yrimmig
language
languaged
speakeng
hcovany
hcovpriv
hinsemp
hinspur
hinstri
hcovpub
hinscaid
hinscare
hinsva
hinsihs
school
educ
educd
gradeatt
gradeattd
schltype
degfield
degfieldd
degfield2
degfield2d
empstat
empstatd
labforce
occ
ind
classwkr
classwkrd
looking
availble
inctot
ftotinc
incwage
incbus00
incss
incwelfr
incinvst
incretir
incsupp
incother
incearn
poverty
migrate1
migrate1d
migplac1
migcounty1
migmet131
vetdisab
diffrem
diffphys
diffmob
diffcare
diffsens
diffeye
diffhear


# Select columns we are interested to look at

In [17]:
sub_df = df[['age', 'inctot', 'educ', 'empstat']].copy()

In [19]:
sub_df.sample(5)

age   inctot                       educ   empstat
119103  68    37200         4 years of college  employed
147984   7  9999999  nursery school to grade 4       n/a
225893  46    52000        5+ years of college  employed
117260  30    20000         2 years of college  employed
311339  54    30000                   grade 12  employed

# Any problems with the data?

In [21]:
sub_df.dtypes

age        category
inctot        int32
educ       category
empstat    category
dtype: object

In [20]:
sub_df.describe()

inctot
count  3.190040e+05
mean   1.723646e+06
std    3.732326e+06
min   -9.000000e+03
25%    1.050000e+04
50%    3.370000e+04
75%    9.140000e+04
max    9.999999e+06

In [33]:
sub_df['inctot'].min()

-9000

In [34]:
sub_df['inctot'].max()

9999999

In [30]:
list(sub_df['age'].unique().sort_values())

['less than 1 year old',
 '1',
 '2',
 '3',
 '4',
 '5',
 '6',
 '7',
 '8',
 '9',
 '10',
 '11',
 '12',
 '13',
 '14',
 '15',
 '16',
 '17',
 '18',
 '19',
 '20',
 '21',
 '22',
 '23',
 '24',
 '25',
 '26',
 '27',
 '28',
 '29',
 '30',
 '31',
 '32',
 '33',
 '34',
 '35',
 '36',
 '37',
 '38',
 '39',
 '40',
 '41',
 '42',
 '43',
 '44',
 '45',
 '46',
 '47',
 '48',
 '49',
 '50',
 '51',
 '52',
 '53',
 '54',
 '55',
 '56',
 '57',
 '58',
 '59',
 '60',
 '61',
 '62',
 '63',
 '64',
 '65',
 '66',
 '67',
 '68',
 '69',
 '70',
 '71',
 '72',
 '73',
 '74',
 '75',
 '76',
 '77',
 '78',
 '79',
 '80',
 '81',
 '82',
 '83',
 '84',
 '85',
 '86',
 '87',
 '88',
 '89',
 '90 (90+ in 1980 and 1990)',
 '91',
 '92',
 '93',
 '94',
 '95',
 '96']

In [31]:
list(sub_df['educ'].unique().sort_values())

['n/a or no schooling',
 'nursery school to grade 4',
 'grade 5, 6, 7, or 8',
 'grade 9',
 'grade 10',
 'grade 11',
 'grade 12',
 '1 year of college',
 '2 years of college',
 '4 years of college',
 '5+ years of college']

In [32]:
list(sub_df['empstat'].unique().sort_values())

['n/a', 'employed', 'unemployed', 'not in labor force']

In [35]:
agg = {'inctot': [min, max]}
sub_df.groupby(['empstat']).agg(agg)

inctot         
                      min      max
empstat                           
n/a                 -2300  9999999
employed            -9000  1563000
unemployed          -7900   566000
not in labor force  -6900   814000

Issues:
- Negative income
- Max income is 9999999, looks to be NA
- Age is represented as categorical, not numerical
- People where empstat is different then employed still have an income value populated

# Clear data

In [72]:
# Let's look only for employed people
cond1 = sub_df['empstat'] == 'employed'
# Let's remove people with negative income
cond2 = sub_df['inctot'] >= 0

filtered_df = sub_df[cond1&cond2].copy()

# Check the data

## Count of categorical variables

In [46]:
filtered_df['age'].value_counts()

54                      3589
53                      3442
56                      3418
55                      3412
47                      3374
                        ... 
5                          0
4                          0
3                          0
2                          0
less than 1 year old       0
Name: age, Length: 97, dtype: int64

In [45]:
filtered_df['educ'].value_counts()

grade 12                     47778
4 years of college           33159
1 year of college            22877
5+ years of college          20985
2 years of college           14066
grade 11                      2744
grade 5, 6, 7, or 8           2091
grade 10                      1904
n/a or no schooling           1290
grade 9                       1290
nursery school to grade 4      468
Name: educ, dtype: int64

## What is the average value of inctot for people whose highest grade level is “grade 12” (in the US, that is someone who has graduated high school)?

In [57]:
cond1 = filtered_df['educ'] <= 'grade 12'
filtered_df.groupby([cond1]).mean()['inctot']

educ
False    71216.312668
True     36827.339460
Name: inctot, dtype: float64

## What is the average income of someone who graduated college (“4 years of college”)? What does that suggest is the value of getting a college degree after graduating high school?

In [53]:
cond1 = filtered_df['educ'] >= '4 years of college'
filtered_df.groupby([cond1]).mean()['inctot']

educ
False    40132.887756
True     88910.444592
Name: inctot, dtype: float64

In [55]:
88910.444592/36827.339460

2.414251094314593

**~2.4 times the salary of someone who has up to Grade 12 level of education**

## What is the average income for someone who has not finished high school? What does that suggest is the value of a high school diploma?

In [54]:
cond1 = filtered_df['educ'] < 'grade 12'
filtered_df.groupby([cond1]).mean()['inctot']

educ
False    60129.100486
True     26261.216818
Name: inctot, dtype: float64

In [58]:
36827.339460/26261.216818

1.4023470319455174

**40% higher than someone who has up to Grade 12 level of education**

## Complete the following table:

Average income for someone who has not finished high school: _________

Average income for someone who only completed 9th grade: _________

Average income for someone who only completed 10th grade: _________

Average income for someone who only completed 11th grade: _________

Average income for someone who finished high school (12th grade) but never started college: _________

Average income for someone who completed 4 year of college (in the US, this means graduating college): _________

In [59]:
cond1 = filtered_df['educ'] < 'grade 12'
filtered_df[cond1]['inctot'].mean()

26261.216818228262

In [60]:
cond1 = filtered_df['educ'] == 'grade 9'
filtered_df[cond1]['inctot'].mean()

27171.907751937986

In [61]:
cond1 = filtered_df['educ'] == 'grade 10'
filtered_df[cond1]['inctot'].mean()

23105.201680672268

In [62]:
cond1 = filtered_df['educ'] == 'grade 11'
filtered_df[cond1]['inctot'].mean()

21570.194606413996

In [63]:
cond1 = filtered_df['educ'] == 'grade 12'
filtered_df[cond1]['inctot'].mean()

38991.73818493867

In [64]:
cond1 = filtered_df['educ'] >= '4 years of college'
filtered_df[cond1]['inctot'].mean()

88910.44459219858

## Why do you think there is no benefit from moving from grade 9 to grade 10, or grade 10 to grade 11, but there is a huge benefit to moving from grade 11 to graduating high school (grade 12)?

In [83]:
filtered_df['f_age'] = filtered_df['age'].cat.codes.astype('float', errors='ignore')

In [87]:
filtered_df.groupby('educ').mean()

inctot      f_age
educ                                               
n/a or no schooling         32306.627907  47.454264
nursery school to grade 4   27592.649573  48.852564
grade 5, 6, 7, or 8         30701.549498  46.530846
grade 9                     27171.907752  41.777519
grade 10                    23105.201681  36.514181
grade 11                    21570.194606  33.462828
grade 12                    38991.738185  43.871070
1 year of college           43169.370022  40.866460
2 years of college          48722.272288  44.148016
4 years of college          75520.718960  43.222383
5+ years of college        110067.933858  46.956016

**It looks like moving from grade 9 to 10, 11 or 12 have little impact on income, in this case the age or years of experience is a better predictor of income.**